weather_mv/loader_pipeline/ee_util.py

```diff
@@ -0,0 +1,457 @@
+
+        # Check that Cloud resource regions are consistent.
+        if not (known_args.dry_run or known_args.skip_region_validation):
+            # Program execution will terminate on failure of region validation.
+            logger.info('Validating regions for data migration. This might take a few seconds...')
+            validate_region(temp_location=pipeline_options_dict.get('temp_location'),
+                            region=pipeline_options_dict.get('region'))
+            logger.info('Region validation completed successfully.')
+
+    def __post_init__(self):
+        """Initializes Sink by initialize earthengine based on user input."""
+
+        if self.dry_run:
+            logger.debug('Uploaded assets into earth engine ...')
+            return
+
+    def expand(self, paths):
+        """Convert gribs from paths into tiff and upload them into earth engine."""
+        (
+            paths
+            | 'ExtractGribData' >> beam.ParDo(
+                ExtractGribData(
+                    open_dataset_kwargs=self.xarray_open_dataset_kwargs,
+                    disable_in_memory_copy=self.disable_in_memory_copy,
+                    disable_grib_schema_normalization=self.disable_grib_schema_normalization))
+            | 'IngestIntoEE' >> beam.ParDo(
+                IngestIntoEE(
+                    dry_run=self.dry_run,
+                    ee_asset=self.ee_asset,
+                    tiff_location=self.tiff_location,
+                    use_personal_account=self.use_personal_account))
+        )
+
+
+@dataclasses.dataclass
+class ExtractGribData(beam.DoFn):
+    """Extract grib data from merged dataset."""
+
+    def __init__(self,
+                 open_dataset_kwargs: t.Optional[t.Dict] = None,
+                 disable_in_memory_copy: bool = False,
+                 disable_grib_schema_normalization: bool = False):
+        self.open_dataset_kwargs = open_dataset_kwargs
+        self.disable_in_memory_copy = disable_in_memory_copy
+        self.disable_grib_schema_normalization = disable_grib_schema_normalization
+
+    def _create_grib_name(self, uri: str) -> str:
+        file_name = os.path.basename(uri)
+        base_name = os.path.splitext(file_name)[0]
+        grib_name = base_name.replace('.', '_')
+        return grib_name
+
+    def process(self, uri: str) -> t.Iterator[GribData]:
+        """Opens grib files and yields channel data."""
+
+        logger.info(f'Converting {uri!r} to COGs ...')
+
+        with open_dataset(uri,
+                          self.open_dataset_kwargs,
+                          self.disable_in_memory_copy,
+                          self.disable_grib_schema_normalization) as ds:
+
+            with open_local(uri) as local_path:
+                with rasterio.open(local_path, 'r') as f:
+                    dtype, crs, transform = (f.profile.get(key) for key in ['dtype', 'crs', 'transform'])
```

Comment:
Do you need to do this? `open_dataset` calls both `open_local` and opens data with rasterio.

Reply:
Maybe, you could refactor `open_dataset` such that `open_local` could be called once? Copying data is an expensive operation (most time is spent in data movement + network io). How can we minimize this cost?

Reply:
Great catch. Thanks.
Moved extracting dtype, crs, projection part to open_dataset() and storing them in resultant ds attrs.